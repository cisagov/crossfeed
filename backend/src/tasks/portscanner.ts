// This is a temporary proof-of-concept port scanner that actively scans ports.
// We'll want to replace this with the Project Sonar passive port scanner.

import { Service } from '../models';
import { plainToClass } from 'class-transformer';
import * as portscanner from 'portscanner';
import getIps from './helpers/getIps';
import { CommandOptions } from './ecs-client';
import saveServicesToDb from './helpers/saveServicesToDb';

export const handler = async (commandOptions: CommandOptions) => {
  const { organizationId, organizationName } = commandOptions;

  console.log('Running portscanner on organization', organizationName);

  const domainsWithIPs = await getIps(organizationId);

  const services: Service[] = [];
  for (const domain of domainsWithIPs) {
    for (const port of [21, 22, 80, 443, 3000, 8080, 8443]) {
      try {
        const status = await portscanner.checkPortStatus(port, domain.ip);
        if (status === 'open') {
          services.push(
            plainToClass(Service, {
              domain: domain,
              discoveredBy: { id: commandOptions.scanId },
              port: port,
              lastSeen: new Date(Date.now())
            })
          );
        }
      } catch (e) {
        console.error(e);
        continue;
      }
    }
  }

  await saveServicesToDb(services);

  console.log(`Portscan finished for ${services.length} services`);
};
