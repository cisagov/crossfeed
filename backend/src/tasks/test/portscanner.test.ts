import { handler as portscanner } from '../portscanner';
jest.mock('../helpers/getIps');
jest.mock('../helpers/saveServicesToDb');

jest.mock('portscanner', () => ({
  checkPortStatus: (port, ip) => {
    return port === 443 || port === 80 ? 'open' : 'closed';
  }
}));

const RealDate = Date;

describe('portscanner', () => {
  beforeEach(() => {
    global.Date.now = jest.fn(() => new Date('2019-04-22T10:20:30Z').getTime());
  });

  afterEach(() => {
    global.Date = RealDate;
  });
  test('basic test', async () => {
    await portscanner({
      organizationId: 'organizationId',
      organizationName: 'organizationName',
      scanId: 'd0f51c16-a64a-4ed0-8373-d66485bfc678',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });
  });
});
