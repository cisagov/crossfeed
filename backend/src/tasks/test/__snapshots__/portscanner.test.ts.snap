// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`portscanner basic test: helpers.saveServicesToDb 1`] = `
Array [
  Service {
    "discoveredBy": Object {
      "id": "d0f51c16-a64a-4ed0-8373-d66485bfc678",
    },
    "domain": Domain {
      "ip": "104.84.119.215",
      "name": "www.cisa.gov",
      "services": Array [
        Object {
          "port": 80,
        },
        Object {
          "port": 443,
        },
      ],
    },
    "lastSeen": 2019-04-22T10:20:30.000Z,
    "port": 80,
  },
  Service {
    "discoveredBy": Object {
      "id": "d0f51c16-a64a-4ed0-8373-d66485bfc678",
    },
    "domain": Domain {
      "ip": "104.84.119.215",
      "name": "www.cisa.gov",
      "services": Array [
        Object {
          "port": 80,
        },
        Object {
          "port": 443,
        },
      ],
    },
    "lastSeen": 2019-04-22T10:20:30.000Z,
    "port": 443,
  },
]
`;
